pkg/cri/resource-manager/requests.go

```diff
@@ -461,16 +466,77 @@ func (m *resmgr) UpdateContainer(ctx context.Context, method string, request int
 	return &criapi.UpdateContainerResourcesResponse{}, nil
 }
 
+type ContainerStats struct {
+	CPU    float64 `containermetrics:"cpu"`
+	Memory float64 `containermetrics:"wss"`
+}
+
+type NumaMetrics struct {
+	Total       float64 `containermetrics:"numa_total"`
+	File        float64 `containermetrics:"numa_file"`
+	Anon        float64 `containermetrics:"numa_anon"`
+	Unevictable float64 `containermetrics:"numa_unevictable"`
+}
+
 // ListContainerStats feeds CRI container statistics to metrics processing.
 func (m *resmgr) ListContainerStats(ctx context.Context, method string, request interface{},
 	handler server.Handler) (interface{}, error) {
 
 	reply, rqerr := handler(ctx, request)
 
+	numaStats := make(map[string]cgroups.NumaStat)
 	filter := request.(*criapi.ListContainerStatsRequest).Filter
 	if filter.Id == "" && filter.PodSandboxId == "" && len(filter.LabelSelector) == 0 {
-		if err := m.SendEvent(reply.(*criapi.ListContainerStatsResponse).Stats); err != nil {
-			m.Error("failed to inject CRI ListContainerStats for event processing: %v", err)
+		cs := make(map[string]interface{})
+		for _, c := range reply.(*criapi.ListContainerStatsResponse).Stats {
+			cs[c.GetAttributes().GetId()] = &ContainerStats{
+				CPU:    float64(c.GetCpu().GetUsageCoreNanoSeconds().GetValue()),
+				Memory: float64(c.GetMemory().GetWorkingSetBytes().GetValue()),
+			}
+
+			ns := utils.GetContainerCgroupDir("/sys/fs/cgroup/memory/kubepods.slice", c.GetAttributes().GetId())
+			if ns == "" {
+				continue
+			}
+			stat, err := cgroups.GetNumaStats(ns)
+			if err != nil {
+				m.Info("failed to get numa stats %+v", err)
+				continue
+			}
+			numaStats[c.GetAttributes().GetId()] = stat
+		}
+		if err := m.SendEvent(&events.Metrics{Data: cs}); err != nil {
+			m.Error("failed to send Container Metrics: %v", err)
+		}
+	}
+
+	// TODO(mythi) Get all NUMA Nodes nicely
+	for jj, ii := range []string{"N0"} {
+		ns := make(map[string]interface{})
+		for cgroup, stat := range numaStats {
+
+			nmt := reflect.TypeOf(NumaMetrics{})
+			stf := []reflect.StructField{}
+
+			// create custom (modify tags) struct type based on NumaMetrics
+			for i := 0; i < nmt.NumField(); i++ {
+				fieldInfo := nmt.Field(i)
+				name := fieldInfo.Tag.Get("containermetrics")
+				fieldInfo.Tag = reflect.StructTag(fmt.Sprintf(`containermetrics:"%s_n%d"`, strings.Trim(name, "\""), jj))
+				stf = append(stf, fieldInfo)
+			}
+
+			v := reflect.New(reflect.StructOf(stf)).Elem()
+
+			v.Field(0).SetFloat(float64(stat.Total.Nodes[ii]))
+			v.Field(1).SetFloat(float64(stat.File.Nodes[ii]))
+			v.Field(2).SetFloat(float64(stat.Anon.Nodes[ii]))
+			v.Field(3).SetFloat(float64(stat.Unevictable.Nodes[ii]))
+
+			ns[cgroup] = v.Addr().Interface()
+		}
+		if err := m.SendEvent(&events.Metrics{Data: ns}); err != nil {
+			m.Error("failed to send NUMA Metrics for node %s: %v", ii, err)
```

Comment:
Wouldn't it be more efficient, easier to read/maintain and also more flexible (read: less inherent implementation-specific restrictions) if the extraction/collection/construction of per-container metrics would be organised into a set of functions, all operating on a pointer to an instance of the 'top-level` per container metrics data structure, which would just be called one-by-one in this loop, each filling in the parts of the structure they are responsible for extracting/collecting/constructing ?

It could also get us rid of this reflect-based code using custom struct tags which tends to be quite expensive.